ANGSD
================
2024-01-17

- [Genotype Likelihoods](#genotype-likelihoods)
  - [Generating a sites file](#generating-a-sites-file)
  - [Generating genotype likelihood
    files](#generating-genotype-likelihood-files)
- [ngsLD pruning](#ngsld-pruning)
  - [Prepare the input files](#prepare-the-input-files)
  - [Run ngsLD](#run-ngsld)
  - [LD pruning](#ld-pruning)
    - [Generate an LD-pruned SNP list](#generate-an-ld-pruned-snp-list)

# Genotype Likelihoods

## Generating a sites file

Before running the genotype likelihood analysis, a sites file should be
prepared. This is a file containing the sites to include in the
analysis. Bed files should be first converted to angsd format, as they
are differently indexed.

To convert a bed file (generated by the GenErode pipeline) to angsd file
and index the file:

``` bash
awk '{print $1"\t"$2+1"\t"$3}' GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.bed > GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.sites
angsd sites index GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.sites
```

Which generates following three files:

- `GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.sites`
- `GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.bin`
- `GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.idx`

In this reference, repeats are masked (“repma”) and only autosomes
(“autos”) are used.

## Generating genotype likelihood files

``` bash
INPUT="Mouflon_domestic"
DIR="/proj/snic2020-2-10/private/Analyses/marianne/PROJECTS/EuropeanMouflon/ANGSD"
INPUTFILE="/proj/snic2020-2-10/private/Analyses/marianne/PROJECTS/EuropeanMouflon/ANGSD/${INPUT}.list"
SITESFILE="/proj/snic2020-2-10/private/Analyses/marianne/PROJECTS/EuropeanMouflon/ANGSD/GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.repma.autos.angsd.file"
REFERENCE="/proj/snic2020-2-10/private/Data/Non-Human/Animals/sheep/ref_seqs/ARS-UI_Ramb_v2.0/RepeatMasker/GCF_016772045.1_ARS-UI_Ramb_v2.0_genomic.fna"

angsd -bam $INPUTFILE \
  -GL 1 \
  -nThreads 16 \
  -doGlf 2 \
  -doMajorMinor 1 \
  -SNP_pval 1e-6 \
  -doMaf 1 \
  -minMapQ 30 -minQ 30 \
  -uniqueOnly 1 -remove_bads 1 \
  -sites $SITESFILE \
    -ref $REFERENCE \
    -out $DIR/$INPUT
```

Which generates following files:

- `Mouflon_domestic.beagle.gz`
- `Mouflon_domestic.mafs.gz`

Parameters:

- `-GL`: Method to estimate genotype likelihood from bam files. Four
  different methods are available.
  - 1.  SAMtools model

  - 2.  GATK model
- `-doGlf`: Output the log genotype likelihoods to a file
  - 0.  Don’t output the genotype likelihoods

  - 1.  Binary all 10 log genotype likelihood

  - 2.  Beagle genotype likelihood format
- `-doMajorMinor`: Infer major and minor allele using a maximum
  likelihood approach ()
  - 1.  Infer major and minor from GL

  - 2.  Infer major and minor from allele counts
- `-SNP_pval 1e-6`: Remove sites with a pvalue larger
- `-doMaf`: Calculate persite frequencies (‘.mafs.gz’)
  - 1.  Frequency (fixed major and minor)
- `-minMapQ`: Minimum mapQ quality.
- `-minQ`: Minimum base quality score.
- `-uniqueOnly`: Remove reads that have multiple best hits. 0 no
  (default), 1 remove.
- `-remove_bads`: Same as the samtools flags -x which removes read with
  a flag above 255 (not primary, failure and duplicate reads). 0 no , 1
  remove (default).
- `-sites`: File containing the sites to include in analysis (chr pos).

#### More resources:

<https://www.popgen.dk/angsd/index.php/Genotype_Likelihoods>  
<https://www.popgen.dk/angsd/index.php/Major_Minor>  
<https://www.popgen.dk/angsd/index.php/Allele_Frequencies>

# ngsLD pruning

Before running PCangsd and NgsAdmixture, the data is first LD-pruned
using the program [ngsLD](https://github.com/fgvieira/ngsLD)

## Prepare the input files

ngsLD requires two input files.

1.  `--geno FILE`: input file with genotypes, genotype likelihoods or
    genotype posterior probabilities. A `beagle` formatted genotype
    likelihood file generated from ANGSD (`-doGlf 2`) can be inputted
    into ngsLD after the header row and the first three columns
    (i.e. positions, major allele, minor allele) are removed.
2.  `--pos FILE`: input file with site coordinates (one per line), where
    the 1st column stands for the chromosome/contig and the 2nd for the
    position (bp). This file is generated by selecting the first two
    columns of the `mafs` file outputted by ANGSD, with the header
    removed. One convenient way to generate this is by selecting the
    first two columns of the `mafs` file outputted by ANGSD, with the
    header removed.

``` bash
#!/bin/bash -l

INPUT="Mouflon_domestic"
INDIR="/proj/snic2020-2-10/private/Analyses/marianne/PROJECTS/EuropeanMouflon/ANGSD"
OUTDIR="/proj/snic2020-2-10/private/Analyses/marianne/PROJECTS/EuropeanMouflon/ngsLD"

## Prepare a geno file
zcat $INDIR/${INPUT}.beagle.gz cut -f 4- | tail -n +2 | gzip  > $OUTDIR/${INPUT}.geno.beagle.gz

## Prepare a pos file
zcat $INDIR/${INPUT}.mafs.gz | cut -f 1,2 | tail -n +2 | gzip > $OUTDIR/${INPUT}.pos.gz
```

## Run ngsLD

``` bash
#!/bin/bash -l

ngsLD \
--geno $OUTDIR/${INPUT}.geno.beagle.gz \
--pos $OUTDIR/${INPUT}.pos.gz \
--probs \
--n_ind 76 \
--n_sites 1134 \
--max_kb_dist 0 \
--n_threads 16 \
--out $OUTDIR/${INPUT}.ld
```

Important ngsLD parameters:

- `--probs`: specification of whether the input is genotype
  probabilities (likelihoods or posteriors)?
- `--n_ind INT`: sample size (number of individuals).
- `--n_sites INT`: total number of sites. To get this information, do
  `bash zcat ${INPUT}.pos.gz | wc -l`.
- `--max_kb_dist DOUBLE`: maximum distance between SNPs (in Kb) to
  calculate LD. Set to 0(zero) to disable filter. \[100\]
- `--max_snp_dist INT`: maximum distance between SNPs (in number of
  SNPs) to calculate LD. Set to 0 (zero) to disable filter. \[0\]
- `--n_threads INT`: number of threads to use. \[1\]
- `--out FILE`: output file name. \[stdout\]

## LD pruning

For many downstream analyses, independence among different SNPs is often
assumed, so it is important to generate a list of SNPs that are in low
LD with each other. To do this, we can use the `prune_graph.pl` script
provided in ngsLD. This script takes the LD estimation output (in our
case `MME_ANGSD_PCA.ld`) as its input. Some important parameters
include:

- `--max_kb_dist INT`: Maximum distance between nodes (ie. SNPs, input
  file 3rd column) to assume they are connected
- `--min_weight FLOAT`: Minimum weight (in –weight_field) of an edge to
  assume nodes are connected (the weight refers to the LD estimate
  between two SNPs)
- `--out FILE`: Path to output file \[STDOUT\]

<br>

``` bash

perl $NGSLD/scripts/prune_graph.pl \
--in_file $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.ld \
--max_kb_dist 2000 \
--min_weight 0.5 \
--out $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled_unlinked.id
```

Check the LD pruning result. How many SNPs survived the LD pruning
process and how many were lost?

<br>

### Generate an LD-pruned SNP list

We will use R to generate an LD-pruned SNP list in a format that can be
used by ANGSD for downstream analyses.

``` r
basedir="~/lcwgs-guide-tutorial/tutorial3_ld_popstructure/"

pruned_position <- as.integer(gsub("Mme_chr24_2558528-4558528:", "", readLines(paste0(basedir, "ngsld/MME_ANGSD_PCA_subsampled_unlinked.id"))))

snp_list <- read.table(paste0(basedir, "angsd/MME_ANGSD_PCA.mafs.gz"), stringsAsFactors = F, header = T)[,1:4]

pruned_snp_list <- snp_list[snp_list$position %in% pruned_position, ]
  
write.table(pruned_snp_list, paste0(basedir, "ngsld/LDpruned_snps.list"), col.names = F, row.names = F, quote = F, sep = "\t")
```
